File: kotlin-app/src/main/kotlin/swimming/service/LLMService.kt
package swimming.service

import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.*

@Serializable
data class ChatMessage(val role: String, val content: String)

@Serializable
data class GeminiResponse(val candidates: List<Candidate> = emptyList())

@Serializable
data class Candidate(val content: GeminiContent = GeminiContent(), val finishReason: String = "")

@Serializable
data class GeminiContent(val parts: List<GeminiPart> = emptyList())

@Serializable
data class GeminiPart(val text: String = "")

class LLMService {

    private val apiKey: String by lazy {
        System.getenv("GEMINI_API_KEY")
            ?: error("GEMINI_API_KEY environment variable is not set. Please set it to a valid Gemini API key.")
    }

    private val model = "gemini-2.5-flash"

    private val client = HttpClient(CIO) {
        install(ContentNegotiation) {
            json(Json { ignoreUnknownKeys = true })
        }
        engine {
            requestTimeout = 60_000
        }
    }

    private val json = Json { ignoreUnknownKeys = true }

    suspend fun chat(systemPrompt: String, messages: List<ChatMessage>, temperature: Double = 0.7): String =
        withContext(Dispatchers.IO) {
            val body = buildJsonObject {
                putJsonObject("systemInstruction") {
                    putJsonArray("parts") {
                        addJsonObject { put("text", systemPrompt) }
                    }
                }
                putJsonArray("contents") {
                    for (msg in messages) {
                        addJsonObject {
                            put("role", if (msg.role == "assistant") "model" else "user")
                            putJsonArray("parts") {
                                addJsonObject { put("text", msg.content) }
                            }
                        }
                    }
                }
                putJsonObject("generationConfig") {
                    put("maxOutputTokens", 8192)
                    put("temperature", temperature) 
                }
            }

            val url = "https://generativelanguage.googleapis.com/v1beta/models/$model:generateContent?key=$apiKey"
            println("[LLM] Calling Gemini model=$model")
            val response = client.post(url) {
                contentType(ContentType.Application.Json)
                setBody(body.toString())
            }

            val responseText = response.bodyAsText()
            println("[LLM] Response status: ${response.status}")
            println("[LLM] Response (first 500 chars): ${responseText.take(500)}")
            val responseJson = json.parseToJsonElement(responseText).jsonObject
            responseJson["error"]?.let { error ->
                val errorMsg = error.jsonObject["message"]?.jsonPrimitive?.content ?: responseText
                throw RuntimeException("Gemini API error: $errorMsg")
            }


            val geminiResponse = json.decodeFromString<GeminiResponse>(responseText)
            if (geminiResponse.candidates.isEmpty()) {
                throw RuntimeException("No candidates in response: $responseText")
            }

            val candidate = geminiResponse.candidates[0]


            if (candidate.finishReason == "MAX_TOKENS") {
                println("LLM WARNING: la respuesta fue truncada (finishReason=MAX_TOKENS)")
            }

            val text = candidate.content.parts.firstOrNull()?.text
                ?: throw RuntimeException("No text in parts: $responseText")
            text
        }
}

File: server/services/LLMService.js
const https = require('https');

const MODEL = 'gemini-2.5-flash';

function getApiKey() {
  const key = process.env.GEMINI_API_KEY;
  if (!key) throw new Error('GEMINI_API_KEY environment variable is not set');
  return key;
}

/**
 * Send a chat request to Gemini.
 * @param {string} systemPrompt
 * @param {Array<{role: string, content: string}>} messages
 * @param {number} temperature
 * @returns {Promise<string>} The model's text response
 */
async function chat(systemPrompt, messages, temperature = 0.7) {
  const apiKey = getApiKey();

  const body = {
    systemInstruction: { parts: [{ text: systemPrompt }] },
    contents: messages.map(msg => ({
      role: msg.role === 'assistant' ? 'model' : 'user',
      parts: [{ text: msg.content }]
    })),
    generationConfig: { maxOutputTokens: 8192, temperature }
  };

  const url = `https://generativelanguage.googleapis.com/v1beta/models/${MODEL}:generateContent?key=${apiKey}`;
  const response = await postJSON(url, body);

  if (response.error) {
    throw new Error(`Gemini API error: ${response.error.message || JSON.stringify(response.error)}`);
  }

  if (!response.candidates || response.candidates.length === 0) {
    throw new Error('No candidates in Gemini response');
  }

  const parts = response.candidates[0].content?.parts;
  if (!parts || parts.length === 0) {
    throw new Error('No text in Gemini response parts');
  }

  return parts[0].text;
}

function postJSON(url, body) {
  return new Promise((resolve, reject) => {
    const parsed = new URL(url);
    const data = JSON.stringify(body);

    const req = https.request({
      hostname: parsed.hostname,
      path: parsed.pathname + parsed.search,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(data)
      },
      timeout: 60000
    }, res => {
      let chunks = '';
      res.on('data', chunk => chunks += chunk);
      res.on('end', () => {
        try { resolve(JSON.parse(chunks)); }
        catch (e) { reject(new Error(`Failed to parse Gemini response: ${chunks.substring(0, 300)}`)); }
      });
    });

    req.on('error', reject);
    req.on('timeout', () => { req.destroy(); reject(new Error('Gemini request timeout (60s)')); });
    req.write(data);
    req.end();
  });
}

module.exports = { chat };
